### PR TITLE
KTOR-3420 Implement multiplatform CLI generator of Ktor projects

### PR DESCRIPTION
**Pull request by Ololoshechkin**
First version of the CLI generator.
For now it only supports windows64, mac64, linux64 and only provides 2 functions:
- `ktor generate <name>`
-`ktor run <name>`

As well as debug version with a custom host:
- `ktor <host> generate <name>`
-`ktor <host> run <name>`

In future some parts may be fixed/changed. Other than that more options will be added.

Note: currently on Windows the only working option that can be run from any shell is the attached script. I've checked that it works under win10 but it still requires some checks under other windows builds. Also this solution may be rewritten to some other tool in the future.

### CODE REVIEW

File: .idea/compiler.xml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (e5l) [inline .idea/compiler.xml]: please add .idea to ignore

File: build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        }
+    }
+    sourceSets {
+        val nativeMain by getting {
+            dependencies {
+                implementation("io.ktor:ktor-client-core:$ktor_version")
+                implementation("org.jetbrains.kotlinx:kotlinx-coroutines-core:1.5.2-native-mt")
+                implementation("io.ktor:ktor-client-serialization:$ktor_version")
+                implementation("com.squareup.okio:okio:3.0.0")
+//                implementation("io.ktor:ktor-client-content-negotiation:1.6.5")
```
Comment (e5l) [inline build.gradle.kts]: comments should be removed

File: src/mingwX64Main/c_interop/WinApi.def [hunk 3]
```diff
@@ -0,0 +1,6 @@
+headers = direct.h
+headerFilter = curl/*
+
+compilerOpts.linux = -I/usr/include -I/usr/include/x86_64-linux-gnu
```
Comment (e5l) [inline src/mingwX64Main/c_interop/WinApi.def:4]: `curl` requires runtime dependency right now, the final binary will not work without additional `dynamic library`

File: src/mingwX64Main/kotlin/io/ktor/generator/cli/CliGenerator.kt [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+import io.ktor.client.*
+import io.ktor.generator.cli.installer.*
+import io.ktor.generator.cli.utils.*
+import okio.FileSystem
+import okio.Path.Companion.toPath
+
+class WindowsKtorInstaller(client: HttpClient, ktorBackendHost: String) : KtorInstaller(client, ktorBackendHost) {
+    override val rootKtorDirName: String = ".ktor."
+    override val jdkDownloadUrl: String =
+        "https://download.java.net/java/GA/jdk11/13/GPL/openjdk-11.0.1_windows-x64_bin.zip"
```
Comment (e5l) [inline src/mingwX64Main/kotlin/io/ktor/generator/cli/CliGenerator.kt]: consider using latest `LTS` jdk version

File: src/nativeMain/kotlin/io/ktor/generator/cli/installer/KtorInstaller.kt [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+
+    private fun findCustomJdk(): Directory? =
+        ktorRootDir
+            .content()
+            .filterIsInstance<Directory>()
+            .find { it.name == JDK_INSTALLED_DIR_PATH }
+            ?.subdir(JAVA_CONTENTS)
+            ?.subdir(JAVA_CONTENTS_HOME)
+
+    private fun hasAdHockJdk(): Boolean = findCustomJdk() != null
```
Comment (e5l) [inline src/nativeMain/kotlin/io/ktor/generator/cli/installer/KtorInstaller.kt]: Consider renaming this method

File: src/nativeMain/kotlin/io/ktor/generator/cli/installer/KtorInstaller.kt [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+        if (hasJdk()) {
+            if (getRcProperty(JAVA_HOME) == null) {
+                addRcProperty(JAVA_HOME, findCustomJdk()?.path ?: getJavaHome()!!)
+            }
+            return
+        }
+
+        val jdkArchiveFile = Directory.home().createFileIfNeeded(jdkArchiveName)
+
+        println("JDK 11 not found.\nJdk download path:$jdkDownloadUrl\nDownloading JDK-11 from server, please wait...")
```
Comment (e5l) [inline src/nativeMain/kotlin/io/ktor/generator/cli/installer/KtorInstaller.kt]: We should extract such string constants in a separate file and consider localization in the future.

It's hard to proofread it right now

File: src/nativeMain/kotlin/io/ktor/generator/cli/installer/KtorInstaller.kt [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+            println("Project $path does not exist")
+            return
+        }
+
+        val gradleFile = projectDir.gradleWrapper() ?: return
+        chdir(projectDir.path)
+        runGradle(gradleFile, GRADLE_RUN, ktorJavaHome, args)
+    }
+
+    protected companion object {
```
Comment (e5l) [inline src/nativeMain/kotlin/io/ktor/generator/cli/installer/KtorInstaller.kt]: can be public

File: src/nativeMain/kotlin/io/ktor/generator/cli/installer/UnixKtorInstaller.kt [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+
+import io.ktor.client.*
+import io.ktor.generator.cli.utils.*
+import okio.FileSystem
+import okio.Path.Companion.toPath
+
+// Common for MacOs and Linux
+abstract class UnixKtorInstaller(
+    osName: String,
+    bitness: String,
```
Comment (e5l) [inline src/nativeMain/kotlin/io/ktor/generator/cli/installer/UnixKtorInstaller.kt]: architecture

File: src/mingwX64Main/kotlin/io/ktor/generator/cli/CliGenerator.kt [hunk 9]
```diff
@@ -0,0 +1,24 @@
+package io.ktor.generator.cli
+
+import io.ktor.client.*
+import io.ktor.generator.cli.installer.*
+import io.ktor.generator.cli.utils.*
+import okio.FileSystem
+import okio.Path.Companion.toPath
+
+class WindowsKtorInstaller(client: HttpClient, ktorBackendHost: String) : KtorInstaller(client, ktorBackendHost) {
```
Comment (e5l) [inline src/mingwX64Main/kotlin/io/ktor/generator/cli/CliGenerator.kt]: I'm not sure about inheritance here, it's not really needed
I would make it as class instances and all logic move to `expect/actual`.

File: src/nativeMain/kotlin/CliGeneratorMain.kt [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+    var argIndex = 0
+    val ktorBackendHost = when {
+        args[0].startsWith(HOST) -> args[argIndex++].removePrefix(HOST)
+        else -> DEFAULT_KTOR_URL
+    }
+
+    println("ktorBackendHost: $ktorBackendHost")
+
+    val command = args.getOrNull(argIndex++) ?: reportError(COMMAND_MISSING)
```
Comment (e5l) [inline src/nativeMain/kotlin/CliGeneratorMain.kt]: We should consider using CLI parsing library instead

File: src/nativeMain/kotlin/io/ktor/generator/cli/installer/KtorInstaller.kt [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+import io.ktor.generator.configuration.json.*
+import io.ktor.http.*
+import kotlinx.cinterop.addressOf
+import kotlinx.cinterop.toKString
+import kotlinx.cinterop.usePinned
+import kotlinx.coroutines.runBlocking
+import platform.posix.*
+import kotlin.system.exitProcess
+
+internal fun getEnv(name: String): String? = getenv(name)?.toKString()
```
Comment (e5l) [inline src/nativeMain/kotlin/io/ktor/generator/cli/installer/KtorInstaller.kt]: Consider extracting utility functions to the separate file

**Review by e5l — CHANGES_REQUESTED** [review]

File: src/mingwX64Main/c_interop/WinApi.def (hunk 3 above)
Comment (Ololoshechkin) [reply src/mingwX64Main/c_interop/WinApi.def:4]: Sorry, forgot to add it in gradle for windows

File: src/mingwX64Main/kotlin/io/ktor/generator/cli/CliGenerator.kt (hunk 4 above)
Comment (Ololoshechkin) [reply src/mingwX64Main/kotlin/io/ktor/generator/cli/CliGenerator.kt]: Why is it important?

**Comment by Stexxe** [comment]
Tests for the core functionality are missing.

**Review by e5l — APPROVED** [review]
Please add a couple of integration tests before merging

**Pull request merged**